### DISCUSSION TITLE
Window resizes incorrectly on startup (Linux/Wayland)

### DISCUSSION THREAD

**Discussion by Kitt3120** [ppy/osu#26241] [Q&A] [answered]
### Type

Game behaviour

### Bug description

When starting up the game on Linux using Wayland and the SDL_VIDEODRIVER=wayland environment variable, the window resizes to an incorrect size.

- It happens on both, the OpenGL (legacy) and the OpenGL renderer.
- It happens on Windowed, Borderless and Fullscreen.
- It happens when using the "Default" resolution and when specifying a resolution manually.

As soon as any of the graphics layout settings are changed, and the game "refreshes", it scales to the correct size.

On my 2560x1440@240 screen, it initially scales to 1365x767, which is about 53% my screen's size.

It hasn't done this 4 days ago. That should've been on 2023.1224.0.

### Screenshots or videos

![image](https://github.com/ppy/osu/assets/10689811/e28d66b0-a51f-4fca-8e75-ce1c71063dc3)

https://github.com/ppy/osu/assets/10689811/e7b8cb45-f5b1-48fb-896b-b213c7f1e075

Sorry, video doesn't seem to play in Firefox.

### Version

2023.1229.1

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/13795954/compressed-logs.zip)

**Comment by Kitt3120** [comment]
This might be something more. I found out that my tablet also doesn't work in osu anymore until I toggle a specific setting on and off again.

To be clear, on Wayland you have to disable "High precision mouse" to get OpenTabletDriver working in osu. This has always been the case and that's okay.

But now, I have to enable and again disable this setting to get my tablet working on each startup.

It seems as if osu starts with some default values and does not consider some saved settings anymore.

This video shows me using my tablet outside of osu (working), when starting up osu (not working) and after toggling "High precision mouse" on and off again (working).


https://github.com/ppy/osu/assets/10689811/bb266775-c13e-4201-a25f-7d251f070daa

**Comment by memesnightmares** [comment]
This is not a wayland issue, I would guess it has to do with your specific DE not wayland in general. I run sway with the same SDL_VIDEODRIVER env var and I don't experience any issues (game works both with/without high precision mouse enabled and setting persists across restarts, tested on 2023.12.29.1). It is something else about your setup. From my understanding of this projects stance on linux, if you want this to be resolved, it may be helpful to go through and find the specific commit/code change that introduced the issue (a good starting point might be before/after https://github.com/ppy/osu/commit/1233533fb967390c2c433ea0f7feb0dadd682635).

**Reply by Kitt3120** [reply]
As there is not "one wayland" (Wayland is just a protocol on paper), it seems that Sway's Wayland handles this correctly and Plasma's Wayland implementation does not.

Good point, I'll try to narrow down the specific commit.

**Reply by Susko3** [reply]
Probably a regression from updating SDL2 in https://github.com/ppy/osu-framework/pull/5309/commits/985922df650c54c956a0e0b6dc21833ffb379f41.

If you're willing, you can try looking for the regression with https://github.com/libsdl-org/SDL/tree/SDL2 itself.

**Reply by Kitt3120** [reply]
Thanks for the tip. I'll look into it if I find the time to do so.

**Reply by THEGOLDENPRO** [reply]
This issue seemed to begin for me after the [2023.1229.1](https://github.com/ppy/osu/releases/tag/2023.1229.1) update.

I'm gonna bring what I said at https://github.com/ppy/osu-framework/pull/6094#issuecomment-1876083099 over here:
> 
> It seems like ever since this release my game has been launching like the image below despite it being in fullscreen mode.
> 
> Toggling fullscreen mode on and off seems to temporarily fix it but it occurs once again when I relaunch.
> 
> ![image](https://github.com/ppy/osu-framework/assets/66202304/38c5fbae-ce67-49d8-86a2-9a5b5c7ff962)
> 
> - OS: Arch Linux x86_64
> - DE: KDE Plasma 5.27.10
> - Kernel: 6.6.9-arch1-1
> 
> Let me know if you need anything else, thank you.

**Comment by Kitt3120** [comment]
Okay, I finally found the time to set the project up on my linux machine. It was a hassle to get dotnet 6 with Roslyn support working. As Arch's packaging of dotnet somehow has Roslyn disabled, I had to use Microsoft's install script, which installs dotnet in user-space and move it into the system dirs manually.

Anyways, I am able to reproduce the error on a fresh build with default settings. Time to bisect!


Edit: Summary of the below conversation

```
osu:
    Last good: b53e7305b4
    First bad: 1233533fb967390c2c433ea0f7feb0dadd682635

Resulting in...
osu-framework:
    Last good: 2023.1227.0
    First bad: 2023.1227.1
    Commits between releases: https://github.com/ppy/osu-framework/compare/2023.1227.0...2023.1227.1

Resulting in...
SDL2-CS:
    Last good: 1.0.671-alpha
    First bad: 1.0.693-alpha
    Commits between releases:
      From: https://github.com/ppy/SDL2-CS/compare/e4afbb848586fca530b6538320f799f81a18b941
      To: https://github.com/ppy/SDL2-CS/commit/c4976a1cc1bcbb64ef6600d157cdf021179a5899
      (Can't get GitHub compare view to work on this)

Resulting in...
Upstream error in SDL-CS wrapper by flibitijibibo (https://github.com/flibitijibibo/SDL2-CS)
or SDL itself (https://github.com/libsdl-org/SDL).

Related SDL issues/PRs in relevant timespan:
  - https://github.com/libsdl-org/SDL/issues/7961
  - https://github.com/libsdl-org/SDL/pull/8013

Which could mean that this is actually expected behaviour from SDL and fixed a possible segfault.
The KDE Plasma Wayland compositor would be to blame here for reporting wrong dimensions.
```

**Reply by Kitt3120** [reply]
```
torben@archeosaurus ~/N/P/D/o/osu.Desktop ((1233533f)|BISECTING) [SIGINT]> git bisect bad
1233533fb967390c2c433ea0f7feb0dadd682635 is the first bad commit
commit 1233533fb967390c2c433ea0f7feb0dadd682635
Author: Dean Herbert <pe@ppy.sh>
Date:   Wed Dec 27 22:14:15 2023 +0900

    Update framework

 osu.Android.props        | 2 +-
 osu.Game/osu.Game.csproj | 2 +-
 osu.iOS.props            | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)
```


I can confirm that [the first bad commit](https://github.com/ppy/osu/commit/1233533fb967390c2c433ea0f7feb0dadd682635) is the one we suspected. Only thing changed in that commit was the osu-framework dependency. Not entirely sure how I can debug that yet, but I'll have a look.

**Reply by Kitt3120** [reply]
Okay, found something interesting. The osu.Framework.Tests project also uses SDL for its window and, when starting with SDL_VIDEODRIVER=wayland, it does not result in the same behaviour. Could mean that it is not a bug in SDL itself, but osu or the osu-framework. Or osu ends up using SDL in such a way that only it triggers the bug.

**Reply by THEGOLDENPRO** [reply]
> Okay, found something interesting. The osu.Framework.Tests project also uses SDL for its window and, when starting with SDL_VIDEODRIVER=wayland, it does not result in the same behaviour. Could mean that it is not a bug in SDL itself, but osu or the osu-framework. Or osu ends up using SDL in such a way that only it triggers the bug.

Since finde but one question, what is SDL?

**Reply by Kitt3120** [reply]
SDL is the cross-platform library osu-framework uses to spawn a window and, I guess, also interact with audio, graphics, etc, as SDL exposes that.

https://github.com/libsdl-org/SDL

**Reply by THEGOLDENPRO** [reply]
> SDL is the cross-platform library osu-framework uses to spawn a window and interact with audio, opengl, etc.

Oh okay, thanks for the explanation. 👍

**Reply by Kitt3120** [reply]
As I can't reproduce the bug using osu-framework's test UI directly, I had to come up with something else. I tried using the older osu-framework version with the latest master commit, but of course that fails, as the newest osu depends on some features of osu-framework that are missing in the old version of osu-framework.

However, using the last good commit, which was [b53e7305b4](https://github.com/ppy/osu/commit/b53e7305b4), I was able to update its osu-framework dependency to the newer 2023.1227.1, and it still compiles and runs. Even better, it now also has the bug when starting on Wayland, further confirming the relation to osu-framework versions > 2023.1219.0. I will now decrease the versions until I can find the specific release of osu-framework that introduced the bug.

**Reply by Kitt3120** [reply]
Well, that was quick. Just one version below, using osu-framework 2023.1227.0, the bug is gone. So it was introduced in 2023.1227.1. Looking at [the commits of that release](https://github.com/ppy/osu-framework/compare/2023.1227.0...2023.1227.1), [the commit you mentioned](https://github.com/ppy/osu-framework/commit/985922df650c54c956a0e0b6dc21833ffb379f41) shows up, @Susko3 

It probably is the SDL update.
The timespan between the last two SDL-CS releases is 7 months. This won't be fun to narrow down to a commit.

**Reply by Kitt3120** [reply]
2023.1227.0 was released on 07.07.2023 and 2023.1227.1 was released on 17.12.2023.
So this includes all commits from [e4afbb8](https://github.com/ppy/SDL2-CS/commit/e4afbb848586fca530b6538320f799f81a18b941) up to [c4976a1](https://github.com/ppy/SDL2-CS/commit/c4976a1cc1bcbb64ef6600d157cdf021179a5899).
Git compare doesn't like it, however. [It shows way too many results](https://github.com/ppy/SDL2-CS/compare/e4afbb848586fca530b6538320f799f81a18b941...c4976a1cc1bcbb64ef6600d157cdf021179a5899).
But basically all commits starting from e4afbb8 up to the latest commit at https://github.com/ppy/SDL2-CS/commits/master/

As all those commits are just updates from upstream, it seems as the bug is indeed in the upstream SDL-CS wrapper or even SDL itself.

Just a guess, but maybe this is already fixed in upstream SDL again and just needs another bump?
I'm not going to dig deep into SDL, but I'll have a look at their issues page.

**Reply by Kitt3120** [reply]
This seems related: https://github.com/libsdl-org/SDL/pull/8013
Could this have possibly introduced this? If so, this is actually expected behaviour from SDL and fixed [a possible segfault](https://github.com/libsdl-org/SDL/issues/7961). The KDE Plasma Wayland compositor would be to blame here for reporting wrong dimensions.

To test this, tomorrow I'll set up VMs with some Plasma and Gnome distros. Maybe Gnome's Wayland compositor behaves differently. Or, if someone wants to help, please go ahead and feel free to do so and report back here :)

**Reply by Susko3** [reply]
For testing this, I suggest you build `libSDL2.so` from source, copy it into the osu! directory (whatever that means on linux) and then run it. At least that's the way I tested it on Windows with copying DLLs. To avoid manually packing SDL2-CS.

You could check [`SDL_GetVersion()`](https://wiki.libsdl.org/SDL2/SDL_GetVersion) to make sure it's using the `libSDL2.so` you gave it.

It's also possible we're doing something slightly wrong in the startup sequence, so the entirety of the fault may not be with SDL. But it would still be useful to track down the offending commit.

**Reply by Kitt3120** [reply]
Hey, it's been a while. Did not have much time to further work on this, and I won't have some time for now. Still, I took a moment to set up a fresh NixOS VM with Plasma 6 and the behaviour is the same. So I am sure that it has nothing to do with my personal setup/environment. Will try out GNOME soon.

**Reply by THEGOLDENPRO** [reply]
yeah sadly this issue hasn't gone away for me, I've just gotten used to re-toggling full-screen. 😅 

Also shouldn't this be converted into an actual issue. @Kitt3120

**Reply by Kitt3120** [reply]
This originally was an issue and has been converted to a conversation. I'll let the moderators here manage this.

As I have a NixOS VM now, it should be easy to switch between Wayland compositors. I'll try GNOME's mutter and other Wayland compositors soon.

**Comment by THEGOLDENPRO** [comment]
If anyone wants a ***lazy*** temporary solution I made a tiny python script that automatically full screens osu! on launch: https://github.com/THEGOLDENPRO/kde-osu-fullscreen-fix

**Comment by Kitt3120** [comment]
It's fixed for me now in 2024-412-1. The game still briefly starts with the wrong dimensions and fixes itself as soon as it gets focused. This also fixed the problems I had with my tablet.

It's worth mentioning that it always worked on my NixOS machine. It was just broken on my Arch installation. Weird. Probably SDL2, maybe I'll investigate it when I have some time. Good thing is that it works for now x)

Edit: Well, my tablet is not really fixed. It works again initially, but it's completely offset to my actual settings and has wrong sensitivity. It's like osu re-maps the tablet input. Tablet input is disabled in osu so it uses my OpenTabletDrivers. It's fixed by, just like before, cycling through windowed, borderless and fullscreen again.

**Comment by Susko3** [comment]
Please give it a try again in the latest version! We've switched to SDL3, so things might have changed.

**Reply by Kitt3120** [reply]
Yup, it's fixed! 🥳🎉 
osu! starts in fullscreen again and I don't have to enable and disable raw input once to get my tablet working.

Keeping this discussion open for another couple of days, so others like @THEGOLDENPRO can give their feedback.

**Reply by THEGOLDENPRO** [reply]
> Keeping this discussion open for another couple of days, so others like @THEGOLDENPRO can give their feedback.

Yeah when I have the time I'll try it out. Glad to hear that.

**Reply by THEGOLDENPRO** [reply]
Just built the flatpak. It seems to fullscreen correctly now. 👍

**Reply by Neon095** [reply]
It's broken again

**Reply by THEGOLDENPRO** [reply]
> It's broken again

what version are you using?

**Reply by Kitt3120** [reply]
I can confirm that 2024.521.2 is still working for me

**Reply by THEGOLDENPRO** [reply]
Just updated to `2024.521.2` and yes @Neon095 is right, it seems to have broken again.

**Reply by Kitt3120** [reply]
Alright. Works for me, but I have reopened the discussion.

**Reply by THEGOLDENPRO** [reply]
Just built and tested `2024.519.1` again and that works, for some reason the latest version (`2024.521.2`) is having the fullscreening problems once again.

**Reply by THEGOLDENPRO** [reply]
> Just built and tested `2024.519.1` again and that works, for some reason the latest version (`2024.521.2`) is having the fullscreening problems once again.

ohh, that's why
![image](https://github.com/ppy/osu/assets/66202304/6bdfb625-0cb5-4f01-93ee-956743c1d7de)

**Reply by Kitt3120** [reply]
You guys made me curious, so I left osu! open for a bit longer. It actually started switching between fullscreen and being glitched like every 5 seconds on its own. Yup, it's not working xD

Okay, the SDL3 revert explains a lot.

**Reply by Kitt3120** [reply]
Good news: osu-framework already has SDL3 re-enabled again! So next release will probably have SDL3 back. For now, you can enable SDL3 by starting osu! with `OSU_SDL3=1` according to [this](https://github.com/ppy/osu-framework/pull/6292).

https://github.com/ppy/osu-framework/releases

Edit:

Can confirm `SDL_VIDEODRIVER=wayland OSU_SDL3=1 osu-lazer` working fine

**Reply by Neon095** [reply]
For me its not working even with the osu sdl3. When i receive any system notification the games goes to windowed mode for a few seconds. This didn't happened before the revert

https://github.com/ppy/osu/assets/97938542/939968d9-b2db-44e3-93e6-483427e5ba8b

For example, this is what happens when i mute the audio

**Reply by Susko3** [reply]
`OSU_SDL3=1` doesn't do anything in the current release. You'll have to wait for the next one (probably).

**Reply by Kitt3120** [reply]
Alright, thanks for clearing up :)

**Comment by Kitt3120** [comment, marked as answer]
https://github.com/ppy/osu/pull/31965 this fixed the mouse problem. No more having to toggle direct input on and off again. SDL also seems to start in full screen and on the correct monitor now. Everything is fixed! 🥳

**Reply by Kitt3120** [reply]
Both problems are back in 2025.225.0 💀  Let's see for how long
